File: 500.600/518.CoinChange2.cpp
#include <iostream>
#include <vector>
using namespace std;

class CoinChange2
{
public:
    int change(int amount, const vector<int> &coins)
    {
        // 此题关键在于巧妙的去重
        // 算了3=2+1 不会计算 3=1+2
        vector<int> dp(amount + 1, 0);
        dp[0] = 1;
        for (int coin : coins)
        {
            for (int i = coin; i <= amount; ++i)
            {
                dp[i] += dp[i - coin];
            }
        }
        return dp[amount];
    }
};

int main()
{
    CoinChange2 cls;
    cout << cls.change(5, vector<int>{1, 2, 5}) << " -> 4" << endl;
    cout << cls.change(3, vector<int>{2}) << " -> 0" << endl;
    cout << cls.change(10, vector<int>{10}) << " -> 1" << endl;
    return 0;
}

File: go/500.600/518.CoinChange2.go
package main

import (
	"fmt"
)

func change(amount int, coins []int) int {
	dp := make([]int, amount+1)
	dp[0] = 1
	for _, v := range coins {
		for i := v; i <= amount; i++ {
			dp[i] += dp[i-v]
		}
	}
	return dp[amount]
}

func main() {
	fmt.Println(change(5, []int{1, 2, 5}), 4)
	fmt.Println(change(3, []int{2}), 0)
	fmt.Println(change(10, []int{10}), 1)
}
